app/parse.py

```diff
@@ -1,12 +1,123 @@
+import logging
+import time
 from dataclasses import dataclass
+from functools import wraps
+from typing import Any
+
+from aiohttp import ClientSession, ClientTimeout
+from bs4 import BeautifulSoup
+import asyncio
+
+BASE_URL = "https://mate.academy/"
 
 
 @dataclass
 class Course:
     name: str
     short_description: str
     duration: str
+    module_count: int
+    topics: list[tuple[str, str]]
+
+
+def configure_logging() -> None:
+    logging.basicConfig(
+        level=logging.DEBUG,
+        format="%(asctime)s - %(levelname)s - %(message)s",
+        handlers=[logging.FileHandler("app.log"), logging.StreamHandler()],
+    )
+
+
+def log_time(func) -> Any:
+    @wraps(func)
+    async def wrapper(*args, **kwargs) -> Any:
+        start_time = time.time()
+        result = await func(*args, **kwargs)
+        logging.info(
+            f"Time taken by {func.__name__}: {time.time() - start_time:.2f} seconds"
+        )
+        return result
+
+    return wrapper
+
+
+async def get_page_content(session: ClientSession, url: str) -> str:
+    try:
+        async with session.get(url, timeout=ClientTimeout(total=5)) as response:
+            response.raise_for_status()
+            return await response.text()
+    except Exception as e:
+        logging.error(f"Error fetching {url}: {e}")
+    return ""
+
+
+def parse_single_course(course: BeautifulSoup) -> Course:
+    modules = course.select("div.CourseModulesList_moduleListItem__HKJqw")
+    topics = [
+        (
+            module.select_one("p.CourseModulesList_topicName__ZrDxT").text,
+            module.select_one(
+                "p.CourseModulesList_topicsCount__yAPxH.typography_landingTextMain__Rc8BD"
+            ).text,
+        )
+        for module in modules
+    ]
+
+    return Course(
+        name=course.select_one("h1[data-qa='profession-title']").text.split(":")[0],
+        short_description=course.select_one("p.typography_landingTextMain__Rc8BD").text,
+        duration=course.select_one(
+            ".ComparisonTable_tableBody__W5hzV.mb-24 > div:nth-of-type(7) > div:nth-of-type(2)"
+        ).text,
+        module_count=len(modules),
+        topics=topics,
+    )
```

Comment:
The `parse_single_course` function assumes that all the selectors will return valid elements. Consider adding checks to ensure that each selector returns a valid element before accessing its `text` attribute. This will prevent potential `AttributeError` exceptions if the HTML structure changes or if some elements are missing.